### PR TITLE
Validate transaction lock times

### LINKED ISSUE #2389: Validate lock times in the transaction verifier
## Motivation

Zebra's transaction verifier needs to check the Transaction `lock_time` field.

### Specifications

The `lock_time` (or `nLockTime`) and `tx_in.sequence` fields are defined in the transaction encoding specification.

https://zips.z.cash/protocol/protocol.pdf#consensusfrombitcoin

> The transaction must be finalized: either its locktime must be in the past (or less than or equal to the current block height), or all of its sequence numbers must be 0xffffffff.

https://developer.bitcoin.org/devguide/transactions.html#non-standard-transactions

Zcash-specific clarification:

> Lock times are only enabled if there is at least one sequence number that is **not** u32::MAX. So transactions with no transparent inputs are always accepted, regardless of the value of the lock time field.

https://github.com/zcash/zips/issues/539

#### Interaction with Input Sequence Numbers

> [There] are four-byte sequence numbers in every input. Sequence numbers were meant to allow multiple signers to agree to update a transaction; when they finished updating the transaction, they could agree to set every input’s sequence number to the four-byte unsigned maximum (0xffffffff), allowing the transaction to be added to a block even if its time lock had not expired.
> 
> Even today, setting all sequence numbers to 0xffffffff (the default in Bitcoin Core) can still disable the time lock, so if you want to use locktime, at least one input must have a sequence number below the maximum. Since sequence numbers are not used by the network for any other purpose, setting any sequence number to zero is sufficient to enable locktime.
>
> Locktime itself is an unsigned 4-byte integer which can be parsed two ways:
> 
> If less than 500 million, locktime is parsed as a block height. The transaction can be added to any block which has this height or higher.
>
> If greater than or equal to 500 million, locktime is parsed using the Unix epoch time format (the number of seconds elapsed since 1970-01-01T00:00 UTC—currently over 1.395 billion). The transaction can be added to any block whose block time is greater than the locktime.

https://developer.bitcoin.org/devguide/transactions.html#locktime-and-sequence-number

Clarification:

> Coinbase inputs have a sequence number, and coinbase transaction lock times are checked just like any other transaction.

https://github.com/zcash/zips/issues/539

#### Interaction with Expiry Height

This rule is in ZIP 203, but we don't think it was ever implemented in `zcashd`:
~~If used in combination with nLockTime, both nLockTime and nExpiryHeight must be block heights.~~

### Designs

Zebra already parses the lock time:
https://github.com/ZcashFoundation/zebra/blob/22f2f06bad7ccc9b47e17212e7c6c8c4a10ab6e3/zebra-chain/src/transaction.rs#L173-L174

https://github.com/ZcashFoundation/zebra/blob/22f2f06bad7ccc9b47e17212e7c6c8c4a10ab6e3/zebra-chain/src/transaction/lock_time.rs#L21-L26

And input sequence numbers:
https://github.com/ZcashFoundation/zebra/blob/22f2f06bad7ccc9b47e17212e7c6c8c4a10ab6e3/zebra-chain/src/transparent.rs#L73-L92

But `lock_time` returns zero lock times as `Height(0)`, rather than `None`.

Similarly, Zebra does not check sequence numbers. If there are no sequence numbers, or they are all `u32::MAX`, `lock_time` should return `None`.

#### Missing Information in Transaction Verifier Request

The block time is not currently sent as part of the transaction verifier request.

### PR DESCRIPTION
**Pull request by jvff**
## Motivation

Transactions can have lock times configured so that they are only included in a block after a certain block height or date and time. This is a consensus rule inherited from Bitcoin and that Zebra must validate.

### Specifications

> The transaction must be finalized: either its locktime must be in the past (or less than or equal to the current block height), or all of its sequence numbers must be 0xffffffff.

https://developer.bitcoin.org/devguide/transactions.html#non-standard-transactions

Zcash-specific clarification:

> Lock times are only enabled if there is at least one sequence number that is **not** u32::MAX. So transactions with no transparent inputs are always accepted, regardless of the value of the lock time field.

https://github.com/zcash/zips/issues/539

#### Interaction with Input Sequence Numbers

> [There] are four-byte sequence numbers in every input. Sequence numbers were meant to allow multiple signers to agree to update a transaction; when they finished updating the transaction, they could agree to set every input’s sequence number to the four-byte unsigned maximum (0xffffffff), allowing the transaction to be added to a block even if its time lock had not expired.
> 
> Even today, setting all sequence numbers to 0xffffffff (the default in Bitcoin Core) can still disable the time lock, so if you want to use locktime, at least one input must have a sequence number below the maximum. Since sequence numbers are not used by the network for any other purpose, setting any sequence number to zero is sufficient to enable locktime.
>
> Locktime itself is an unsigned 4-byte integer which can be parsed two ways:
> 
> If less than 500 million, locktime is parsed as a block height. The transaction can be added to any block which has this height or higher.
>
> If greater than or equal to 500 million, locktime is parsed using the Unix epoch time format (the number of seconds elapsed since 1970-01-01T00:00 UTC—currently over 1.395 billion). The transaction can be added to any block whose block time is greater than the locktime.

https://developer.bitcoin.org/devguide/transactions.html#locktime-and-sequence-number

Clarification:

> Coinbase inputs have a sequence number, and coinbase transaction lock times are checked just like any other transaction.

https://github.com/zcash/zips/issues/539

## Solution

- Validate if the sequence numbers enable the transaction's lock time in `Transaction::lock_time` getter
  - If the lock time is disabled (or zero), `None` is returned
- Validate the lock time in a new function in the `zebra_consensus::transaction::check` module
  - That function is called by the transaction verifier on all transactions

## Review

@teor2345 and/or @oxarbitrage can review this. This is still in draft because I have only outlined the tests I want to add, but I haven't finished writing them yet.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

### CODE REVIEW

File: zebra-chain/src/transaction.rs
Comment (teor2345) [inline zebra-chain/src/transaction.rs]: Required Consensus Rule: This looks like redundant code, but it's not. Let's document that.

```suggestion
        // `zcashd` checks that the block height is greater than the lock height.
        // This check allows the genesis block transaction, which would otherwise be invalid.
        // (Or have to use a lock time.)
        //
        // It matches the `zcashd` check here:
        // https://github.com/zcash/zcash/blob/1a7c2a3b04bcad6549be6d571bfdff8af9a2c814/src/main.cpp#L720
        if lock_time == LockTime::unlocked() {
            return None;
        }
```

File: zebra-chain/src/transaction.rs
Comment (teor2345) [inline zebra-chain/src/transaction.rs]: Required Consensus rule clarifications:

1. Bitcoin doesn't allow transactions with no transparent inputs. This is Zcash-specific extension to transactions, so we need to explicitly follow `zcashd` here.

2. Coinbase inputs have a sequence number.

Optional:

3. Let's add a summary sentence without any negations.

```suggestion
        // In `zcashd`, this rule applies to both coinbase and prevout input sequence numbers.
        //
        // Unlike Bitcoin, Zcash allows transactions with no transparent inputs. These transactions only have shielded inputs.
        // Surprisingly, the `zcashd` implementation ignores the lock time in these transactions.
        // `zcashd` only checks the lock time when it finds a transparent input sequence number that is
        // not `u32::MAX`.
```

File: zebra-chain/src/transaction.rs
Comment (teor2345) [inline zebra-chain/src/transaction.rs]: Optional Refactor: This code seems more explicit.

I also moved the comment that explains what the code does, so it's right next to the code.

```suggestion

        // Only return the lock time if a transparent sequence number enables it.
        if self
            .inputs()
            .iter()
            .map(transparent::Input::sequence)
            .any(|sequence_number| *sequence_number != u32::MAX) {
                Some(lock_time)
        } else {
                None
        }

```

File: zebra-consensus/src/transaction/check.rs
Comment (teor2345) [inline zebra-consensus/src/transaction/check.rs]: Optional Naming: The comparisons are easier to read if the block data is explicitly named.

```suggestion
pub fn lock_time_has_passed(tx: &Transaction, block_height: Height, block_time: _) -> Result<(), TransactionError> {
```

File: zebra-consensus/src/transaction/check.rs
Comment (teor2345) [inline zebra-consensus/src/transaction/check.rs]: Required Consensus Rule Change:

1. Consensus rules should not use the local node's time. Node clocks can be set to different times, which would cause a consensus split. (As an exception, we check the block time against the node's local time, plus a 2 hour leeway for clock skew.)

Here is Zebra's block time:
https://github.com/ZcashFoundation/zebra/blob/43c6d013a7b120ccdcfcfa15c3de2e76b3c99853/zebra-chain/src/block/header.rs#L61-L63

2. The Bitcoin time docs say "greater than", not "greater than or equal to".

3. The Bitcoin height docs are different to what `zcashd` actually does. The block height/time must be less than the transaction lock height/time:
https://github.com/zcash/zcash/blob/1a7c2a3b04bcad6549be6d571bfdff8af9a2c814/src/main.cpp#L722

This means the Bitcoin doc is wrong for height, but correct for time:

> The transaction can be added to any block which has ~~this height or~~ higher.
> The transaction can be added to any block whose block time is greater than the locktime.

https://developer.bitcoin.org/devguide/transactions.html#locktime-and-sequence-number

File: zebra-consensus/src/error.rs
Comment (teor2345) [inline zebra-consensus/src/error.rs]: Required Consensus Rule Clarifications:

1. this is a block time, not a user's clock time
2. the lock still applies if the block time is equal to this time

```suggestion
    #[error("transaction is locked until after block time {0}")]
    #[cfg_attr(any(test, feature = "proptest-impl"), proptest(skip))]
    LockedUntilAfterDateTime(DateTime<Utc>),
```

File: zebra-consensus/src/transaction/check.rs
Comment (teor2345) [inline zebra-consensus/src/transaction/check.rs]: Optional Refactor: We're more likely to get the rules correct if they're next to the code that implements them, and the code follows the rule.

```suggestion
            // > The transaction can be added to any block which has a greater height.
            // The Bitcoin documentation is wrong or outdated here,
            // so this code is based on the `zcashd` implementation at:
            // https://github.com/zcash/zcash/blob/1a7c2a3b04bcad6549be6d571bfdff8af9a2c814/src/main.cpp#L722
            if block_height > unlock_height {
                return Ok(());
            } else {
                return Err(TransactionError::LockedUntilBlockHeight(unlock_height));
            }
        }
        Some(LockTime::Time(unlock_time)) => {
            // > The transaction can be added to any block whose block time is greater than the locktime.
            // https://developer.bitcoin.org/devguide/transactions.html#locktime-and-sequence-number
            if block_time > unlock_time {
                return Ok(());
            } else {
```

File: zebra-consensus/src/error.rs
Comment (teor2345) [inline zebra-consensus/src/error.rs]: Required Consensus Rule Clarification: The lock still applies if the block height is equal to this height.

```suggestion
    #[error("transaction is locked until after block height {}", _0.0)]
    LockedUntilAfterBlockHeight(block::Height),
```

**Review by teor2345 — COMMENTED** [review]
I think the Bitcoin docs are wrong about the height check.

Also, we need to be careful to use the correct time field and comparison operators.

File: zebra-consensus/src/transaction/tests.rs
Comment (teor2345) [inline zebra-consensus/src/transaction/tests.rs]: Likely Test Bug: This change creates extra test failures. If the block time is the minimum block time, there is no valid unlock time, because the unlock time must be less than the block time.

```suggestion
                time: chrono::MAX_DATETIME,
```

File: zebra-consensus/src/transaction/tests/prop.rs
Comment (teor2345) [inline zebra-consensus/src/transaction/tests/prop.rs]: PR Scope: I don't think we need to have proptests to merge this PR, particularly if they're going to take a few days. It also seems like some of the proptest strategies are missing from this PR.

Can you split the proptests into a separate PR, so we can schedule that work after NU5 validation is complete?

As a quick test, we should make sure that all transactions in the test vectors pass the new check.
(We can include any proptests that are ready to merge, but please don't spend much time on them.)

File: zebra-consensus/src/transaction.rs
Comment (teor2345) [inline zebra-consensus/src/transaction.rs]: PR Scope: Since the mempool it is not part of the consensus rules, any mempool changes are optional.

Can we split the mempool changes into a separate PR, and skip any new rules for mempool transactions in this PR? (See my previous detailed comment.)

File: zebra-chain/src/transaction.rs
Comment (jvff) [reply zebra-chain/src/transaction.rs]: Done in [8da0661](https://github.com/jvff/zebra/commit/8da06612d22319e32bf23825696398f567638a61), but I didn't add the extra comment because I thought that the binding name would be enough, but I can add it back if you think it's better.

File: zebra-consensus/src/transaction/check.rs
Comment (jvff) [reply zebra-consensus/src/transaction/check.rs]: Done in [7d77ba4](https://github.com/jvff/zebra/commit/7d77ba428faf3fbdd27f493843d1e0c9b160f4b6).

File: zebra-consensus/src/transaction/check.rs
Comment (jvff) [reply zebra-consensus/src/transaction/check.rs]: Done through the following commits (oldest first):

- [fb6efe7](https://github.com/jvff/zebra/commit/fb6efe720ccfd98962359248c2c480080de2db30).
- [833af3e](https://github.com/jvff/zebra/commit/833af3e1ae5b749b02dd7a68ca43037f6a031798).
- [1f900fb](https://github.com/jvff/zebra/commit/1f900fbe7341235ee7b245273301a402ff7c5ae4).
- [55e88e6](https://github.com/jvff/zebra/commit/55e88e640881cea446ae9b15629fd43fdff3f6bb).

Difference with all four commits together can be seen [here](https://github.com/jvff/zebra/compare/fb6efe720ccfd98962359248c2c480080de2db30%5E..55e88e640881cea446ae9b15629fd43fdff3f6bb).

File: zebra-consensus/src/error.rs
Comment (jvff) [reply zebra-consensus/src/error.rs]: Done in [b15fb1e](https://github.com/jvff/zebra/commit/b15fb1efcd86611316d48d1c8d55a8f7a0f826c0) and the new names are used in [85df78b](https://github.com/jvff/zebra/commit/85df78b8d3c3de508460f89dfa36d6833467a6a2).

File: zebra-consensus/src/transaction/check.rs
Comment (jvff) [reply zebra-consensus/src/transaction/check.rs]: Done in [85df78b](https://github.com/jvff/zebra/commit/85df78b8d3c3de508460f89dfa36d6833467a6a2) and [944c684](https://github.com/jvff/zebra/commit/944c68466b8e7c55ad390077be48f07f8ddc9c67). There's also [dc6fabd](https://github.com/jvff/zebra/commit/dc6fabd03bddb418383786b44688450b098d1629)

**Comment by jvff** [comment]
Updated. Difference since last review can be seen [here](https://github.com/jvff/zebra/compare/validate-transaction-lock-times-r1..validate-transaction-lock-times-r1-fixes).

I still haven't finished the tests, I'll try to do that on Monday.

I'm not sure how to handle the mempool transactions. For the block height, I think it's currently using the current tip height plus one:

https://github.com/ZcashFoundation/zebra/blob/0960e4fb0b155009dcf68018e0a51df0df84d3ff/zebrad/src/components/mempool/downloads.rs#L296

Should we add a tolerance of a few extra blocks?

As for the block time, I think the options are:

- only unlocked transactions are accepted in the mempool, 
- locked transactions are accepted and stay in the mempool until they are unlocked (however long that may take),
- only transactions that will be unlocked in at most a future time are accepted in the mempool.

The current implementation uses the third option using the local node time plus a two hours margin of error between the local node time and the network's block time, plus a tolerance of an hour.  Should I use the state service in the mempool downloader to get the block time and pass it in the request instead of using the local time in the verifier itself?

**Comment by teor2345** [comment]
Let's try to limit the scope of this PR to essential NU5 tasks.
We can schedule extra tests or mempool changes in future sprints.

> I'm not sure how to handle the mempool transactions. For the block height, I think it's currently using the current tip height plus one:
> https://github.com/ZcashFoundation/zebra/blob/0960e4fb0b155009dcf68018e0a51df0df84d3ff/zebrad/src/components/mempool/downloads.rs#L296

Can we split mempool validation into a separate PR?
It's not required for NU5.

The mempool and lock time rules are completely undocumented for Zcash. The mempool is not consensus-critical, but we want to avoid usability bugs where Zebra rejects transactions that `zcashd` accepts. So it is better to skip mempool rules.

In that PR, we should consider:
* What's the goal here?
* What rules are we actually implementing?

I think `zcashd` uses the block height and median time past of the current chain tip. But we need to check. I really want to avoid using the local node's time, because expecting users to have an accurate clock is a usability issue.

**Review by teor2345 — COMMENTED** [review]
I'm concerned that the scope of this PR has become a lot larger.

Can we just do the NU5 changes and basic test vectors in the initial PR?
Then we can schedule other work in future sprints.

File: zebra-consensus/src/transaction/tests.rs
Comment (jvff) [reply zebra-consensus/src/transaction/tests.rs]: Done in [e0ad6e4](https://github.com/jvff/zebra/commit/e0ad6e4ce03afd998ae5703eeab39f72d2e18ec4).

File: zebra-consensus/src/transaction/tests/prop.rs
Comment (jvff) [reply zebra-consensus/src/transaction/tests/prop.rs]: Done. I removed the stubbed property tests in [fe97323](https://github.com/jvff/zebra/commit/fe973238447db7a5215626a04c91f179545786cc).

I'll open a separate PR for the property tests shortly.

File: zebra-consensus/src/transaction.rs
Comment (jvff) [reply zebra-consensus/src/transaction.rs]: Done, I refactored it to only validate the lock time for transactions in blocks in [3f41b33](https://github.com/jvff/zebra/commit/3f41b339c251f76b3944825703a5f8b121f2e600) and [5f2d4f4](https://github.com/jvff/zebra/commit/5f2d4f46ebec6b06a5387b60f5018d9c72012a36).

**Comment by jvff** [comment]
Updated. Since I removed the incomplete property tests, I think this PR is no longer blocked by it so can have its draft status removed.

Changes since last review can be seen [here](https://github.com/jvff/zebra/compare/validate-transaction-lock-times-r2..validate-transaction-lock-times-r2-fixes).

**Review by teor2345 — DISMISSED** [review]
Thanks, looks great!

I'm just about to fix up the unrelated coverage job failure.

**Review by teor2345 — APPROVED** [review]
I merged the new errors from two different PRs.

**Pull request merged**
